File: solutions/python/dnd-character/1/dnd_character.py
import random

def modifier(ability):
    return (ability - 10) // 2

class Character:
    __die = [1, 2, 3, 4, 5, 6]
    def __roll(self):
        dice = []
        for i in range(4):
            dice.append(random.choice(self.__die))
        dice.remove(min(dice))
        return sum(dice)
    def __init__(self):
        self.strength = self.__roll()
        self.dexterity = self.__roll()
        self.constitution = self.__roll()
        self.intelligence = self.__roll()
        self.wisdom = self.__roll()
        self.charisma = self.__roll()
        self.hitpoints = 10 + modifier(self.constitution)

    def ability(self):
        return random.choice([self.strength, self.dexterity, self.constitution, self.intelligence, self.wisdom, self.charisma])

File: solutions/fsharp/dnd-character/1/DndCharacter.fs
module DndCharacter

open System

let modifier x = 
    if x < 10 && x % 2 = 1 then
        (x - 10) / 2 - 1
    else
        (x - 10) / 2

let ability() = 
    let dice = new Random()
    let four = [1..4] |> List.map (fun x -> dice.Next(1, 6)) |> List.sort
    match four with
    | small::rest -> rest
    | [] -> []
    |> List.sum
    
type Character = {
    Strength: int
    Dexterity: int
    Constitution: int
    Intelligence: int
    Wisdom: int
    Charisma: int
    Hitpoints: int
}

let createCharacter() =
    let character = {
        Strength = ability()
        Dexterity = ability()
        Constitution = ability()
        Intelligence = ability()
        Wisdom = ability()
        Charisma = ability()
        Hitpoints = 0
    }
    { character with Hitpoints = 10 + modifier character.Constitution }
